### ISSUE TITLE
FP16 with forward only and FP32 in backward

### ISSUE THREAD

**Issue by YuShen1116** [pytorch/pytorch#63719]
### Please note that this issue tracker is not a help form and this issue will be closed.

We have a set of [listed resources available on the website](https://pytorch.org/resources). Our primary means of support is our discussion forum:

- [Discussion Forum](https://discuss.pytorch.org/)

Hi Pyotrch team,

I'm trying to do an experiment with FP16 forwarding and FP32 loss computation/backward and gradient updates,
but now Pytorch has some issues on that when loss is very big. It seems the model's gradient is already half precision even after I manually change it to float(). Is there anything I can do to avoid this problem? Many Thanks!

Below are my test script, the print at the end will show all grad are `Nan`, but with autocast = False, grad won't be `Nan`, 

```python
import torch
from torch import nn
from torch.cuda.amp import autocast

torch.manual_seed(42)


class Dummy(nn.Module):

    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(1, 3, 3)
        self.linear = nn.Linear(3, 1)

    def forward(self, x):
        y = self.conv(x)
        y = y.view(-1)
        y = self.linear(y)
        return y


scaler = torch.cuda.amp.GradScaler(init_scale=1e-15)
dummy_img = torch.rand((1, 1, 3, 3), dtype=torch.float).cuda()
gt = torch.rand(1, dtype=torch.float).cuda()
loss = nn.MSELoss()
model = Dummy().cuda()

with autocast(enabled=True):
    output = model(dummy_img)

model = model.float()
output = output.float()
print("gt, output", gt.type(), output.type())
_loss = loss(gt, output) * 1e15
print("loss", _loss.type())
# print(_loss, scaler.scale(_loss))
# _loss = scaler.scale(_loss)
_loss.backward()
for idx, p in enumerate(model.parameters()):
    print("layer{0} grad".format(idx))
    print(p.grad.type())
    print(p.grad)

```

**Comment by bdhirsh** [comment]
Hi Yu!

I think your question would be better suited for the PyTorch forum. Do you mind re-posting your question there? I'm going to close the issue.

Forum link: https://discuss.pytorch.org/
